File: C++/14-Backtracking/Problems/3-CombinationSum/CombinationSum.js
/**
 * Five different approaches to solve the Combination Sum problem using backtracking in JavaScript.
 *
 * Problem:
 * Given an array of distinct integers `candidates` and a target integer `target`,
 * return a list of all unique combinations of `candidates` where the chosen numbers sum to `target`.
 * You may assume that the same number can be chosen from `candidates` an unlimited number of times.
 * Two combinations are unique if the frequency of at least one of the chosen numbers is different.
 *
 * Example:
 * Input: candidates = [2,3,6,7], target = 7
 * Output: [[2,2,3],[7]]
 */

/**
 * Approach 1: Basic Backtracking
 * - Explore all possible combinations using recursion.
 * - If the current sum equals the target, add the combination to the result.
 * - If the current sum exceeds the target, backtrack.
 */
function combinationSum1(candidates, target) {
  const result = [];
  const combination = [];

  function backtrack(remaining, start) {
    if (remaining === 0) {
      result.push([...combination]);
      return;
    }
    if (remaining < 0) {
      return;
    }

    for (let i = start; i < candidates.length; i++) {
      combination.push(candidates[i]);
      backtrack(remaining - candidates[i], i); // Allow reuse of the same element
      combination.pop(); // Backtrack: remove the last element to try a different path
    }
  }

  backtrack(target, 0);
  return result;
}

/**
 * Approach 2: Backtracking with Early Pruning
 * - Optimized backtracking by adding a condition to avoid unnecessary recursion.
 * - Sort the candidates array to potentially prune earlier.  This doesn't change the fundamental time complexity, but can improve performance in some cases.
 */
function combinationSum2(candidates, target) {
    const result = [];
    const combination = [];
    candidates.sort((a, b) => a - b); // Sort candidates to potentially prune earlier

    function backtrack(remaining, start) {
        if (remaining === 0) {
            result.push([...combination]);
            return;
        }
        if (remaining < 0) {
            return;
        }

        for (let i = start; i < candidates.length; i++) {
            // Early pruning: if current candidate is larger than remaining, skip it
            if (candidates[i] > remaining) {
                break; // Since candidates are sorted, no need to check further
            }
            combination.push(candidates[i]);
            backtrack(remaining - candidates[i], i);
            combination.pop();
        }
    }

    backtrack(target, 0);
    return result;
}


/**
 * Approach 3: Backtracking with Optimized Sum Calculation
 * - Avoids repeated subtraction by calculating the sum incrementally.
 */
function combinationSum3(candidates, target) {
  const result = [];
  const combination = [];
  let currentSum = 0;

  function backtrack(start) {
    if (currentSum === target) {
      result.push([...combination]);
      return;
    }
    if (currentSum > target) {
      return;
    }

    for (let i = start; i < candidates.length; i++) {
      combination.push(candidates[i]);
      currentSum += candidates[i];
      backtrack(i);
      currentSum -= candidates[i]; // Backtrack: subtract the added value
      combination.pop();
    }
  }

  backtrack(0);
  return result;
}

/**
 * Approach 4: Iterative Backtracking (using a stack)
 * - Simulates the recursive backtracking process using a stack.
 * - More complex to understand but avoids recursion overhead.
 */
function combinationSum4(candidates, target) {
  const result = [];
  const stack = [{ combination: [], remaining: target, start: 0 }];

  while (stack.length > 0) {
    const { combination, remaining, start } = stack.pop();

    if (remaining === 0) {
      result.push([...combination]);
    } else if (remaining > 0) {
      for (let i = start; i < candidates.length; i++) {
        stack.push({
          combination: [...combination, candidates[i]],
          remaining: remaining - candidates[i],
          start: i,
        });
      }
    }
    // If remaining < 0, we do nothing (implicit backtracking)
  }
  return result;
}

/**
 * Approach 5: Backtracking with Memoization (Dynamic Programming - Top Down)
 * - Uses memoization to store intermediate results and avoid redundant calculations.
 * - This approach is generally NOT recommended for the *combination sum* problem when unlimited repetitions are allowed, as it doesn't provide significant performance benefits and adds complexity.  Memoization is more helpful when each candidate can be used only *once*.  It is included here for completeness and to show how memoization *could* be applied, but its efficiency is not great here.
 */
function combinationSum5(candidates, target) {
    const memo = new Map(); // Map to store results for subproblems
    const result = [];

    function backtrack(remaining, start) {
        if (remaining === 0) {
            return [[]]; // Base case: one empty combination
        }
        if (remaining < 0) {
            return [];
        }

        const key = `${remaining}-${start}`; // Unique key for the subproblem
        if (memo.has(key)) {
            return memo.get(key);
        }

        const combinations = [];
        for (let i = start; i < candidates.length; i++) {
            const subCombinations = backtrack(remaining - candidates[i], i);
            for (const subCombination of subCombinations) {
                combinations.push([candidates[i], ...subCombination]);
            }
        }
        memo.set(key, combinations);
        return combinations;
    }
    const finalCombinations = backtrack(target, 0);

    // Convert the result from array of arrays to the desired format.
    for(const comb of finalCombinations){
        result.push(comb);
    }
    return result;
}



// Example usage and output:
const candidates = [2, 3, 6, 7];
const target = 7;

console.log("Approach 1: Basic Backtracking");
console.log(combinationSum1(candidates, target)); // Output: [[2,2,3],[7]]

console.log("\nApproach 2: Backtracking with Early Pruning");
console.log(combinationSum2(candidates, target)); // Output: [[2,2,3],[7]]

console.log("\nApproach 3: Backtracking with Optimized Sum Calculation");
console.log(combinationSum3(candidates, target)); // Output: [[2,2,3],[7]]

console.log("\nApproach 4: Iterative Backtracking (using a stack)");
console.log(combinationSum4(candidates, target)); // Output: [[2,2,3],[7]]

console.log("\nApproach 5: Backtracking with Memoization (Dynamic Programming - Top Down)");
console.log(combinationSum5(candidates, target)); // Output: [[2,2,3],[7]]

File: C++/14-Backtracking/Problems/3-CombinationSum/CombinationSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric> // for std::accumulate

using namespace std;

// Approach 1: Basic Backtracking
// - Uses recursion to explore all possible combinations.
// - Includes a base case to stop recursion when the target sum is reached.
// - Time Complexity: O(2^n) in the worst case, where n is the number of candidates.
// - Space Complexity: O(n) for the depth of the recursion tree.
void combinationSum1(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
    if (target == 0) {
        result.push_back(combination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        combination.push_back(candidates[i]);
        combinationSum1(candidates, target - candidates[i], result, combination, i); // Allow duplicates
        combination.pop_back(); // Backtrack
    }
}

// Approach 2: Backtracking with Pruning
// - Optimizes the basic backtracking approach by adding a pruning step.
// - Sorts the candidates array to avoid unnecessary recursive calls.
// - If a candidate is greater than the remaining target, the branch is pruned.
// - Time Complexity: O(b^k), where b is the average branching factor and k is the maximum depth of the recursion.
// - Space Complexity: O(k), where k is the maximum depth of the recursion.
void combinationSum2(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
    if (target == 0) {
        result.push_back(combination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        if (candidates[i] > target) { // Pruning step: if current candidate is too big, skip it and the rest
            break; // Assuming candidates are sorted
        }
        combination.push_back(candidates[i]);
        combinationSum2(candidates, target - candidates[i], result, combination, i);
        combination.pop_back();
    }
}

// Approach 3: Backtracking with Skipping Duplicates
// - Handles duplicate candidates in the input array.
// - Skips duplicate candidates to avoid duplicate combinations in the result.
// - Time Complexity: Similar to Approach 2, but avoids redundant computations.
// - Space Complexity: O(k)
void combinationSum3(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
    if (target == 0) {
        result.push_back(combination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        if (i > start && candidates[i] == candidates[i - 1]) { // Skip duplicates
            continue;
        }
        if (candidates[i] > target) {
            break;
        }
        combination.push_back(candidates[i]);
        combinationSum3(candidates, target - candidates[i], result, combination, i);
        combination.pop_back();
    }
}

// Approach 4: Iterative Approach using Stack (Simulated Recursion)
// - Simulates the recursive backtracking using a stack.
// - Avoids recursion overhead, but can be more complex to understand.
// - Time Complexity: Similar to the recursive approaches.
// - Space Complexity: O(k), where k is the maximum depth of the combinations.
void combinationSum4(vector<int>& candidates, int target, vector<vector<int>>& result) {
    sort(candidates.begin(), candidates.end()); // Need to sort for optimization
    vector<pair<int, vector<int>>> stack; // Store remaining target and current combination
    stack.push_back({target, {}});

    while (!stack.empty()) {
        int remainingTarget = stack.back().first;
        vector<int> combination = stack.back().second;
        stack.pop_back();

        if (remainingTarget == 0) {
            result.push_back(combination);
            continue;
        }

        for (int i = 0; i < candidates.size(); ++i) {
            if (candidates[i] <= remainingTarget) {
                vector<int> nextCombination = combination;
                nextCombination.push_back(candidates[i]);
                stack.push_back({remainingTarget - candidates[i], nextCombination});
            }
        }
    }
}

// Approach 5: Dynamic Programming (Unbounded Knapsack Variation)
// - Treats the problem as an unbounded knapsack problem.
// - dp[i] stores all combinations that sum up to i.
// - Time Complexity: O(target * n), where n is the number of candidates.
// - Space Complexity: O(target * k), where k is the average size of the combinations.

void combinationSum5(vector<int>& candidates, int target, vector<vector<int>>& result) {
    vector<vector<vector<int>>> dp(target + 1); // dp[i] stores list of combinations that sum to i

    dp[0].push_back({}); // Base case: sum 0 has an empty combination

    for (int i = 1; i <= target; ++i) {
        for (int j = 0; j < candidates.size(); ++j) {
            if (i - candidates[j] >= 0) {
                for (const auto& prevCombination : dp[i - candidates[j]]) {
                    vector<int> currentCombination = prevCombination;
                    currentCombination.push_back(candidates[j]);
                    dp[i].push_back(currentCombination);
                }
            }
        }
    }
    result = dp[target];
}

int main() {
    vector<int> candidates = {2, 3, 6, 7};
    int target = 7;
    vector<vector<int>> result;

    cout << "Candidates: ";
    for (int c : candidates) {
        cout << c << " ";
    }
    cout << endl;
    cout << "Target: " << target << endl << endl;

    // Approach 1: Basic Backtracking
    result.clear();
    vector<int> combination1;
    combinationSum1(candidates, target, result, combination1, 0);
    cout << "Approach 1: Basic Backtracking" << endl;
    for (const auto& comb : result) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Approach 2: Backtracking with Pruning
    result.clear();
    vector<int> combination2;
    sort(candidates.begin(), candidates.end()); // Important for pruning
    combinationSum2(candidates, target, result, combination2, 0);
    cout << "Approach 2: Backtracking with Pruning" << endl;
    for (const auto& comb : result) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Approach 3: Backtracking with Skipping Duplicates
    result.clear();
    vector<int> combination3;
    vector<int> candidates3 = {10, 1, 2, 7, 6, 1, 5}; // Example with duplicates
    sort(candidates3.begin(), candidates3.end());
    combinationSum3(candidates3, 8, result, combination3, 0);
    cout << "Approach 3: Backtracking with Skipping Duplicates" << endl;
    for (const auto& comb : result) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Approach 4: Iterative Approach using Stack
    result.clear();
    combinationSum4(candidates, target, result);
    cout << "Approach 4: Iterative Approach using Stack" << endl;
    for (const auto& comb : result) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

     // Approach 5: Dynamic Programming
    result.clear();
    combinationSum5(candidates, target, result);
    cout << "Approach 5: Dynamic Programming" << endl;
    for (const auto& comb : result) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    return 0;
}
